fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,135 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => this.setState({ todos: data }));
+    // .then(data => console.log(data))
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    const data = newTodo;
+    this.setState(state => ({
+      todos: state.todos.concat([newTodo])
+    }));
```

Comment:
This should be in your 2nd .then. You don't want to update state until you have a response from the server. 2 reasons for this: 1. Your new todo won't have an id, the id is generated by the database, and 2. you don't want to show the user that something happened by displaying it on the screen like it was added, then the request fail. If that happens you give the user a false impression that everything worked. So you don't want to update the DOM until the server sends back a response indicating that everything worked out.